src/metabase/driver/query_processor.clj

```diff
@@ -115,17 +127,32 @@
            qp))))
 
 
-(defn- pre-expand [qp]
+(defn- pre-expand-macros
+  "Looks for macros in a structured (unexpanded) query and substitutes the macros for their contents."
+  [qp]
   (fn [query]
-    (qp (if (structured-query? query)
-          (let [macro-expanded-query (macros/expand-macros query)]
-            (when (and (not *disable-qp-logging*)
-                       (not= macro-expanded-query query))
-              (log/debug (u/format-color 'cyan "\n\nMACRO/SUBSTITUTED: 😻\n%s" (u/pprint-to-str macro-expanded-query))))
-            (-> macro-expanded-query
-                expand/expand
-                resolve/resolve))
-          query))))
+    ;; if necessary, handle macro substitution
+    (let [query (if-not (structured-query? query)
+                  query
+                  ;; for structured queries run our macro expansion
+                  (u/prog1 (macros/expand-macros query)
+                    (when (and (not *disable-qp-logging*)
+                               (not= <> query))
+                      (log/debug (u/format-color 'cyan "\n\nMACRO/SUBSTITUTED: 😻\n%s" (u/pprint-to-str <>))))))]
+      (qp query))))
```

Comment:
the macro stuff should ideally be moved into `expand`, that would simplify the impl a bit.